Use kCTF as a virtual environment

sroettger opened the pull request:
This is a big change in the way kCTF works.

You install it by copying the contents from the `dist` directory into your challenge directory.

To use it, you run source kctf/activate. This
* downloads dependencies (currently yq)
* installs a bash function `kctf`
* creates a temporary gcloud + kube config
* sets up a bunch of environment variables
* cleans up on exit

kctf will forward commands to scripts in kctf/bin/

Cluster configurations have names now and live in kctf/config.
You can create/load them with `kctf config`.

stephenR committed:
add venv activate script

stephenR committed:
venv activate + config create/load implementation

stephenR committed:
move cluster-start to dist dir

stephenR committed:
Makefile in dist

stephenR committed:
download yq and handle setup errors

stephenR committed:
challenge create in dist

stephenR committed:
cluster stop in dist

stephenR committed:
remove bin folder

stephenR committed:
remove old scripts

stephenR committed:
cluster resize in dist

stephenR committed:
fix usage and add required args

stephenR committed:
unify usage messages

stephenR committed:
don't use source for kctf-config

stephenR committed:
add deactivate command

stephenR committed:
automatically load last config

stephenR committed:
proper if

stephenR committed:
exit instead of return

stephenR committed:
print usage to stderr everywhere

stephenR committed:
remove some debugging leftover

sirdarckcat reviewed (commented):
looks good so far.. do you want to push updates on this PR? or on new ones.

sroettger commented:
Probably on this one. Since I need to update the workflow and if I change
the interface i need to update it again.

On Tue, Dec 29, 2020, 18:35 sirdarckcat <notifications@github.com> wrote:

> *@sirdarckcat* commented on this pull request.
>
> looks good so far.. do you want to push updates on this PR? or on new ones.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/kctf/pull/180#pullrequestreview-559633131>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEM2JHQHLQQR6DSJ73MO4R3SXIHNVANCNFSM4VNHGJ4A>
> .
>

stephenR committed:
implement config list

stephenR committed:
implement chal list

stephenR committed:
challenge start/stop implementation

stephenR committed:
implement chal status

stephenR committed:
export the ctf name

stephenR committed:
implement chal debug commands

stephenR committed:
remove Makefile

stephenR committed:
update workflow

stephenR committed:
automatic update all the sample images

stephenR committed:
Automated commit: update images.

stephenR committed:
bash completion for the kctf command

stephenR committed:
add debug output to workflow

stephenR committed:
remove --challenge-name flag add --help to all commands

stephenR committed:
update healthcheck as well in workflow

stephenR committed:
challenge in prompt

sroettger commented:
If the tests pass, then let's merge this.
I added:
* bash completion
* challenge name from cwd only
* challenge name in prompt
* automatic updating of the sample images
* implemented all the kctf commands in bash instead of Makefile

sroettger commented:
Next things I want to add:
* update docs
* add support back for local clusters
* kctf.cloud
* fix the load balancer issues
* turn the samples into templates

I think all of these make sense as separate PRs now that this one works and (hoepfully) passes all tests.

stephenR committed:
Automated commit: update images.

sirdarckcat commented on .github/workflows/update-images.yaml:
```diff
@@ -172,41 +170,54 @@ jobs:
+              echo "deployment/${CHALLENGE_NAME} doesn't exist yet, sleeping"
+              sleep 5
+            done
+            kubectl wait --for=condition=available --timeout=5m "deployment/${CHALLENGE_NAME}"
```
whats the difference between this and the for loop above? I assume the answer is that `kubectl wait` fails if the deployment doesnt exist, but it would be nice to add a comment if thats the expectation here

sroettger replied:
hm? Isn't that pretty much what the comment already says?:

```
# We want to wait for the deployment to be available, but it
# might not have been created yet by the operator.
```

sirdarckcat replied:
I mean, that this code is relying on the fact that kubectl wait will fail if the deployment hasn't been created after 5 times 5 seconds.

sroettger replied:
Hmm, I thought I replied to this already. I added some text to the comment.

sirdarckcat commented on dist/activate:
```diff
@@ -0,0 +1,171 @@
+# limitations under the License.
+
+# set up the environment variables
+KCTF_CONFIG_DIR="${HOME}/.config/kctf"
```
do we still need this? make `mktemp -d` is better?

sroettger replied:
done

sirdarckcat commented on dist/activate:
```diff
@@ -0,0 +1,171 @@
+    if [[ -e "${KCTF_BIN}/yq" ]]; then
+      rm "${KCTF_BIN}/yq" >/dev/null
+    fi
+    wget "https://github.com/mikefarah/yq/releases/download/v4.2.0/yq_linux_amd64" -O "${KCTF_BIN}/yq" --quiet || return 1
```
maybe this and the hash below can be env vars at the top.

sroettger replied:
done

sirdarckcat commented on dist/activate:
```diff
@@ -0,0 +1,171 @@
+}
+
+function _kctf_usage {
+  echo -e "usage: kctf command subcommand [args]" >&2
```
nit:
```bash
cat >&2 <<EOF
Usage: kctf command subcommand [args]

Available commands:
  chal:     commands for managing challenges
                    (create, start, stop, debug, ...)
  cluster:  commands for managing the cluster
                    (create, start, stop, resize, ...)
  config:   commands for managing the kctf config
                    (create, load)
EOF
```

sroettger replied:
That way I have to left align everything to the beginning of the line don't I?

sirdarckcat replied:
yep, it's a styling thing - in python this is the usual way to do it (with `"""..."""`)

sroettger replied:
In my opinion it looks pretty bad since it doesn't follow the indentation of the code block. I.e. I would use it only for unindented code

sirdarckcat commented on dist/activate:
```diff
@@ -0,0 +1,171 @@
+          return 1
+        fi
+        source "${KCTF_CTF_DIR}/kctf/config/${CONFIG_NAME}"
+        export PROJECT
```
should this be a `_kctf_export_gcp_vars` like the other exports?

sroettger replied:
I removed the other export function

sirdarckcat commented on dist/activate:
```diff
@@ -0,0 +1,171 @@
+
+function _kctf_error_cleanup {
+  unset -f _kctf_download_dependencies
+  #unset -f _kctf_cleanup
```
?

sroettger replied:
I think I didn't figure out how to remove the trap, so this function will get called on exit

sirdarckcat replied:
I meant, the `#` commented line :)

sroettger replied:
I know, I'm talking about _kctf_cleanup. Added a comment

sirdarckcat commented on dist/activate:
```diff
@@ -0,0 +1,171 @@
+  source "${KCTF_BIN}/kctf-completion"
+}
+
+function _kctf_error_cleanup {
```
is there a better way to accomplish this? maybe we can create a new shell and run there? like

```bash
(echo 'KCTF_ACTIVATE=1 source ' $0; cat) | bash
```

or something like that

sroettger replied:
you mean so that we don't have to unset the variables?

sirdarckcat replied:
so we dont forget to unset one, yep

sroettger replied:
It's not really an issue if we forget to unset a variable. It's mostly to be nice to not leave clutter in the shell.

sirdarckcat commented on dist/activate:
```diff
@@ -0,0 +1,171 @@
+# Copyright 2020 Google LLC
```
nit: `dist` usually is a generated directory, and in this case it is the source directory. this might be confusing to others

sroettger replied:
do you have a suggestion for a better name?
I understand dist as distribute and these are the files that will get distributed in the kctf.tar.gz, that's why I chose the name.

sirdarckcat replied:
maybe sdk?

sroettger replied:
is it an sdk though? I.e. you don't use this to develop software, you use it for deployment.
Maybe it's a software deployment kit, which we could abbreviate as "sdk" ;)
I mean, in the end it doesn't matter too much since the users don't see this.

sirdarckcat replied:
maybe client? idk :P it's just a styling thing

sroettger replied:
let's decide later and change in a future PR?

sirdarckcat replied:
yea :) - the nits is my abbreviation for "let's not bikesched this, but in case you agree, what about"

stephenR committed:
use tmpdir instead of ~/.config/kctf and add yq vars

stephenR committed:
fix the logic to check for changes to commit

stephenR committed:
some comments

stephenR committed:
Automated commit: update images.
